server/decoders/notebook.ts

```diff
@@ -0,0 +1,23 @@
+// ---------------------------------------------------------------------------------------------------------------------
+// Notebook Decoders
+// ---------------------------------------------------------------------------------------------------------------------
+
+import * as JsonDecoder from 'decoders';
+
+// ---------------------------------------------------------------------------------------------------------------------
+
+export const notebookPageDecoder = JsonDecoder.object({
+    id: JsonDecoder.optional(JsonDecoder.either(JsonDecoder.string, JsonDecoder.number)),
+    title: JsonDecoder.string,
+    content: JsonDecoder.string,
+    notebookID: JsonDecoder.string
+});
+
+// ---------------------------------------------------------------------------------------------------------------------
+
+export const notebookDecoder = JsonDecoder.object({
+    id: JsonDecoder.string,
+    pages: JsonDecoder.optional(JsonDecoder.array(notebookPageDecoder))
```

Comment:
Maybe the optional should have an default of `[]`? Make the notebook always have the pages property be an array?

Reply:
I agree, but I can't figure out how to do defaults with decoders? Did I miss something?

Reply:
There doesn't seem to be a built-in way to do that, so there's a couple options.

We can use a map to turn an undefined into an empty array:
```javascript
JsonDecoder.map(JsonDecoder.optional(JsonDecoder.array(notebookPageDecoder)), v => { return v || []; } )
```

We could also use `either` with a `hardcoded` value. This will end up accepting any input though, valid or not, and turn it into an empty array.
```javascript
JsonDecoder.either(JsonDecoder.array(notebookPageDecoder), JsonDecoder.hardcoded([]))
```

Reply:
This was resolved in later commits.